### PR TITLE
Feature/validation timeout

### LINKED ISSUE #251: Wrap critical validation methods in future
As discussed in PR #205 some validation methods like the use of Regex might have a unforeseen long computation time, which can be utilized for denial of service attacks. The discussed solution is to wrap such critical computations in a `Future` and execute it with a reasonable timeout. This automated approach would also avoid unpredictable worst cases.

### PR DESCRIPTION
**Pull request by lrschaef**
### Reason for this PR
- #251 

### Changes in this PR
- Closes #251 

- [X] Test written
- [x] Changelog updated

### CODE REVIEW

File: product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala [hunk 1]
```diff
@@ -51,19 +51,35 @@ class AuthenticationServiceImpl(readSide: ReadSide, processor: AuthenticationEve
...
             throw UC4Exception.OwnerMismatch
           }
-          if (role != user.role) {
+
+          val validationErrors = try {
+            Await.result(authUser.validate, 5.seconds)
+          }
+          catch {
+            case _: TimeoutException => throw UC4Exception.ValidationTimeout
+            case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala:69]: This needs to be changed into a Internal Server Error
```suggestion
            case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

File: product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala [hunk 2]
```diff
@@ -65,7 +65,15 @@ class CourseServiceImpl(
...
           }
 
-          val validationErrors = courseProposal.validate
+          // For syntax and regex checks, 5 seconds are more than enough
+          val validationErrors = try {
+            Await.result(courseProposal.validate, 5.seconds)
+          }
+          catch {
+            case _: TimeoutException => throw UC4Exception.ValidationTimeout
+            case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala:74]: This needs to be changed into a Internal Server Error
```suggestion
            case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

File: product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala [hunk 3]
```diff
@@ -140,7 +148,14 @@ class CourseServiceImpl(
           throw UC4Exception.PathParameterMismatch
         }
 
-        val validationErrors = updatedCourse.validate
+        val validationErrors = try {
+          Await.result(updatedCourse.validate, 5.seconds)
+        }
+        catch {
+          case _: TimeoutException => throw UC4Exception.ValidationTimeout
+          case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala:156]: This needs to be changed into a Internal Server Error
```suggestion
            case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

File: product_code/matriculation_service/impl/src/main/scala/de/upb/cs/uc4/matriculation/impl/MatriculationServiceImpl.scala [hunk 4]
```diff
@@ -41,7 +41,16 @@ class MatriculationServiceImpl(clusterSharding: ClusterSharding, userService: Us
...
         val message = rawMessage.trim
-        val validationList = message.validate
+
+        // For syntax and regex checks, 5 seconds are more than enough
+        val validationList = try {
+          Await.result(message.validate, 5.seconds)
+        }
+        catch {
+          case _: TimeoutException => throw UC4Exception.ValidationTimeout
+          case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/matriculation_service/impl/src/main/scala/de/upb/cs/uc4/matriculation/impl/MatriculationServiceImpl.scala:51]: This needs to be changed into a Internal Server Error
```suggestion
          case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

File: product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala [hunk 5]
```diff
@@ -95,62 +95,73 @@ class UserServiceImpl(
...
 
-      validationErrorsFuture.flatMap {
-        valErrors =>
-          var validationErrors = valErrors
+      var validationErrors = try {
+        Await.result(validationErrorsFuture, 5.seconds)
+      }
+      catch {
+        case _: TimeoutException => throw UC4Exception.ValidationTimeout
+        case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala:122]: This needs to be changed into a Internal Server Error
```suggestion
        case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

File: product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala [hunk 6]
```diff
@@ -160,74 +171,84 @@ class UserServiceImpl(
...
+
+            //validate new user and check, if username errors exist, since entityRef might fail if username is incorrect
+            var validationErrors = try {
+              Await.result(user.validate, 5.seconds)
             }
-            if (editErrors.nonEmpty) {
-              throw new UC4Exception(422, DetailedError(ErrorType.UneditableFields, editErrors))
+            catch {
+              case _: TimeoutException => throw UC4Exception.ValidationTimeout
+              case e: Exception        => throw e
```
Comment (david-buderus) [inline product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala:193]: This needs to be changed into a Internal Server Error
```suggestion
              case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```

**Review by david-buderus — CHANGES_REQUESTED** [review]

**Review by ldklenner — APPROVED** [review]

File: product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala [hunk 7]
```diff
@@ -51,19 +51,35 @@ class AuthenticationServiceImpl(readSide: ReadSide, processor: AuthenticationEve
...
             throw UC4Exception.PathParameterMismatch
           }
-          if (authUsername != user.username.trim) {
+          if (authUsername != authUser.username) {
             throw UC4Exception.OwnerMismatch
           }
-          if (role != user.role) {
+
+          val validationErrors = try {
+            Await.result(authUser.validate, 5.seconds)
```
Comment (david-buderus) [inline product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala:65]: The timeout should be defined in the application.conf
And than loaded into a lazy val

File: product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala [hunk 8]
```diff
@@ -140,7 +148,14 @@ class CourseServiceImpl(
           throw UC4Exception.PathParameterMismatch
         }
 
-        val validationErrors = updatedCourse.validate
+        val validationErrors = try {
+          Await.result(updatedCourse.validate, 5.seconds)
```
Comment (david-buderus) [inline product_code/course_service/impl/src/main/scala/de/upb/cs/uc4/course/impl/CourseServiceImpl.scala:152]: The timeout should be defined in the application.conf
And than loaded into a lazy val

File: product_code/matriculation_service/impl/src/main/scala/de/upb/cs/uc4/matriculation/impl/MatriculationServiceImpl.scala [hunk 9]
```diff
@@ -41,7 +41,16 @@ class MatriculationServiceImpl(clusterSharding: ClusterSharding, userService: Us
     authenticated[PutMessageMatriculation, Done](AuthenticationRole.Admin) {
       ServerServiceCall { (header, rawMessage) =>
         val message = rawMessage.trim
-        val validationList = message.validate
+
+        // For syntax and regex checks, 5 seconds are more than enough
+        val validationList = try {
+          Await.result(message.validate, 5.seconds)
```
Comment (david-buderus) [inline product_code/matriculation_service/impl/src/main/scala/de/upb/cs/uc4/matriculation/impl/MatriculationServiceImpl.scala:47]: The timeout should be defined in the application.conf
And than loaded into a lazy val

File: product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala [hunk 10]
```diff
@@ -95,62 +95,73 @@ class UserServiceImpl(
...
-          Future.successful(postMessageUser.validate)
+          //For other users we cannot encounter duplicate matriculationIDs
+          postMessageUser.validate
       }
 
-      validationErrorsFuture.flatMap {
-        valErrors =>
-          var validationErrors = valErrors
+      var validationErrors = try {
+        Await.result(validationErrorsFuture, 5.seconds)
```
Comment (david-buderus) [inline product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala:118]: The timeout should be defined in the application.conf
And than loaded into a lazy val

File: product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala [hunk 11]
```diff
@@ -160,74 +171,84 @@ class UserServiceImpl(
...
             }
-            val oldUser = optUser.get
-            // Other users than admins can only edit specified fields
-            var editErrors = oldUser.checkUneditableFields(user)
-            if (role != AuthenticationRole.Admin) {
-              editErrors ++= oldUser.checkProtectedFields(user)
+
+            //validate new user and check, if username errors exist, since entityRef might fail if username is incorrect
+            var validationErrors = try {
+              Await.result(user.validate, 5.seconds)
```
Comment (david-buderus) [inline product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala:189]: The timeout should be defined in the application.conf
And than loaded into a lazy val

**Review by david-buderus — CHANGES_REQUESTED** [review]

**Review by david-buderus — APPROVED** [review]

**Pull request merged**